pczt/src/transparent.rs

```diff
@@ -0,0 +1,411 @@
+
+        // Leverage the early-exit behaviour of zip to confirm that the remaining data in
+        // the other bundle matches this one.
+        for (lhs, rhs) in self.inputs.iter_mut().zip(inputs.into_iter()) {
+            // Destructure `rhs` to ensure we handle everything.
+            let Input {
+                prevout_txid,
+                prevout_index,
+                sequence,
+                required_time_lock_time,
+                required_height_lock_time,
+                script_sig,
+                value,
+                script_pubkey,
+                redeem_script,
+                partial_signatures,
+                sighash_type,
+                bip32_derivation,
+                ripemd160_preimages,
+                sha256_preimages,
+                hash160_preimages,
+                hash256_preimages,
+                proprietary,
+            } = rhs;
+
+            if lhs.prevout_txid != prevout_txid
+                || lhs.prevout_index != prevout_index
+                || lhs.value != value
+                || lhs.script_pubkey != script_pubkey
+                || lhs.sighash_type != sighash_type
+            {
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.sequence, sequence)
+                && merge_optional(&mut lhs.required_time_lock_time, required_time_lock_time)
+                && merge_optional(
+                    &mut lhs.required_height_lock_time,
+                    required_height_lock_time,
+                )
+                && merge_optional(&mut lhs.script_sig, script_sig)
+                && merge_optional(&mut lhs.redeem_script, redeem_script)
+                && merge_map(&mut lhs.partial_signatures, partial_signatures)
+                && merge_map(&mut lhs.bip32_derivation, bip32_derivation)
+                && merge_map(&mut lhs.ripemd160_preimages, ripemd160_preimages)
+                && merge_map(&mut lhs.sha256_preimages, sha256_preimages)
+                && merge_map(&mut lhs.hash160_preimages, hash160_preimages)
+                && merge_map(&mut lhs.hash256_preimages, hash256_preimages)
+                && merge_map(&mut lhs.proprietary, proprietary))
+            {
+                return None;
+            }
+        }
+
+        for (lhs, rhs) in self.outputs.iter_mut().zip(outputs.into_iter()) {
+            // Destructure `rhs` to ensure we handle everything.
+            let Output {
+                value,
+                script_pubkey,
+                redeem_script,
+                bip32_derivation,
+                proprietary,
+            } = rhs;
+
+            if lhs.value != value || lhs.script_pubkey != script_pubkey {
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.redeem_script, redeem_script)
+                && merge_map(&mut lhs.bip32_derivation, bip32_derivation)
+                && merge_map(&mut lhs.proprietary, proprietary))
```

Comment:
Here we can assume that if any of the fields on the LHS were `None`, then they were not needed to compute any sighash, and so there is no possibility of a merge invalidating a signature. It might be useful to point that out in a comment.